doc/click-install.1
.\" -*- mode: nroff -*-
.ds V 1.2
.ds D 27/Apr/2000
.ds E " \-\- 
.if t .ds E \(em
.de Sp
.if n .sp
.if t .sp 0.4
..
.de Es
.Sp
.RS 5
.nf
..
.de Ee
.fi
.RE
.PP
..
.de Rs
.RS
.Sp
..
.de Re
.Sp
.RE
..
.de M
.BR "\\$1" "(\\$2)\\$3"
..
.de RM
.RB "\\$1" "\\$2" "(\\$3)\\$4"
..
.TH CLICK-INSTALL 1 "\*D" "Version \*V"
.SH NAME
click-install \- installs a Click configuration into the kernel module driver
'
.SH SYNOPSIS
.B click-install
.RI \%[ options ]
.RI \%[ param = value " ...]"
.RI \%[ router\-file ]
'
.SH DESCRIPTION
.B Click-install
is a convenience program that installs a Click router configuration
into the current Linux kernel. It reads a router configuration file in the
.M click 5
language, installs the
.M click.o 8
kernel module if necessary, compiles and/or installs any dynamically
loadable packages required by the router configuration, installs the router
configuration itself by writing it to /click/config, and uninstalls
any unused packages. It reports any errors to standard error, including
errors reported by the Click module, and returns a zero (true) exit status
if and only if the configuration was successfully installed.
'
.SH "OPTIONS"
'
Arguments such as "NAME=value" set configuration parameters, overriding any
existing settings. Thus, "echo \%'Message($MSG)' | \fBclick\-install\fR
MSG=Hi" will cause the kernel module to print "Hi".
.PP
If any filename argument is a single dash "-",
.B click-install
will use the standard input or output instead, as appropriate.
'
.TP 5
.BI \-f " file"
.PD 0
.TP
.BI \-\-file " file"
Read the router configuration to transform from
.IR file .
The default is the standard input.
'
.Sp
.TP
.BI \-e " expr"
.TP
.BI \-\-expression " expr"
Use
.IR expr ,
a string in the Click language, as the router configuration.
'
.Sp
.TP 5
.BR \-h ", " \-\-hotswap
Hot-swap install the new configuration. If you supply this option, the new
configuration will only be installed if it initializes correctly. If it
fails, the old configuration will continue to route packets without
interruption. Furthermore, state from the old configuration\*Esuch as
packets contained in
.M Queue n
elements\*Ewill be swapped into the new configuration under certain
conditions. See under
.B /click/hotconfig
in
.M click.o 8
for details.
'
.Sp
.TP 5
.BR \-u ", " \-\-uninstall
Uninstall and reinstall the Click kernel module.
.B \-\-hotswap
and
.B \-\-uninstall
are mutually exclusive.
'
.Sp
.TP 5
.BR \-j ", " "\-\-threads=\fIN\fP"
The Click kernel module will run
.I N
threads per router. Click must have been build with
\f(CW\-\-enable\-multithread\fR. This option only matters when
.B click-install
installs a Click kernel module. The thread count on an installed kernel
module cannot be changed. (The
.B \-\-uninstall
option is useful here.)
'
.Sp
.TP 5
.BR \-n ", " "\-\-priority=\fIN\fP"
Set the Click kernel thread's priority to
.IR N .
Lower numbers have higher priority. A typical range for priority values is
-20 (highest) to +19 (lowest). If you do not give a
.B \-\-priority
option, then the current priority is left unchanged.
'
.Sp
.TP 5
.BR \-G ", " "\-\-greedy"
The Click kernel module will claim an entire CPU, not allowing any other
process to run.  This option only matters when
.B click-install
installs a Click kernel module; the greediness of an installed kernel
module cannot be changed.  (The
.B \-\-uninstall
option is useful here.)
'
.Sp
.TP 5
.BR \-p ", " "\-\-private"
Make the /click directory hierarchy readable only by the owning user, which
is usually root. Generally, files under /click are writable only by the
owner, but readable by everyone. Works only when simultaneously loading a
Click kernel module; see
.BR \-\-uninstall .
'
.Sp
.TP 5
.BR \-U ", " "\-\-user=\fIUSER\fR[:\fIGROUP\fR]"
Set the user ID and group ID of the /click directory hierarchy to
.I USER
and group
.IR GROUP .
By default
.IR USER " and " GROUP
are both root. Works only when simultaneously loading a
Click kernel module; see
.BR \-\-uninstall .
'
.Sp
.TP 5
.BR \-\-cpu =\fIN
Sets the Click thread to preferentially run on CPU
.IR N .
If Click is running
.I K
threads, they are run preferentially on CPUs
.I N
up to, but not including,
.I N
+
.IR K .
Thread affinities are ignored if they refer to unavailable CPUs. Works
only when simultaneously loading a Click kernel module; see
.BR \-\-uninstall .
'
.Sp
.TP 5
.BR \-m ", " \-\-map
(Linux 2.4 only) If a Click kernel module is installed, write its load map
to standard output. This can make it easier to debug kernel panics.
'
.Sp
.TP 5
.BR \-V ", " \-\-verbose
Print information about each step in the install process, including the
filenames of any installed modules.
'
.Sp
.TP 5
.BI \-\-help
Print usage information and exit.
'
.Sp
.TP
.BI \-\-version
Print the version number and some quickie warranty information and exit.
'
.PD
'
.SH "EXIT STATUS"
.B Click-install
exits with one of the following return values:
.TP 5
.B 0
The configuration was successfully installed.
.TP
.B 1
Some error prevented
.B click-install
from attempting to install the configuration. For example, perhaps the
configuration could not be parsed, or the kernel module could not be
loaded.
.TP
.B 2
The configuration was installed, but the kernel module reports that it
could not be initialized.
.PD
'
.SH "SEE ALSO"
.M click-uninstall 1 ,
.M click 5 ,
.M click.o 8
'
.SH AUTHOR
.na
Eddie Kohler, kohler@seas.harvard.edu
.br
http://www.pdos.lcs.mit.edu/click/
'


NEWS
Click NEWS

Version 2.0.1   24.Sep.2011

* ns-3 support, thanks to Lalith Suresh and Sascha Alexander Jopen.

* Important bug fixes for user-level ToDevice thanks to Jaeyong Yoo.

* Better handling for time: Timestamp knows about steady clocks (also known
  as monotonic clocks -- clocks that never go backwards) on most platforms,
  so Timers should go off at predictable intervals even when the system
  time is set backwards.

* IPRewriter and friends: TCP supports >2GB transfers and multiple sequence
  number transitions; UDP supports different timeouts for streaming
  connections and RPC-like connections.

* Configure with --enable-stats=2 for more precise cycle counting.

* Classifier and friends (IPFilter, etc.): Correct bugs and improve
  optimization.

* Bug fixes for AddressInfo, user-level packet pool (thanks to Seiichi
  Tetsukawa).

* Support packages on patchless Linux installs.

* Thanks to Cliff Frey, and to Felipe Chaulet, Sascha Alexander Jopen,
  Christian Kreibich, Meraki, Inc., Roberto Riggio, Jim Roewe, Robert
  Sombrutzki, Lalith Suresh, Dimitris Syrivelis, Seiichi Tetsukawa, Michael
  W.S. Yee, Jaeyong Yoo, and all other contributors.


Version 2.0   16.Jul.2011

* New elements: AdjustTimestamp, BlockThread, ClickyInfo, EtherVLANEncap,
  HandlerTask, ICMPIPEncap, NullTask, SetVLANAnno, StoreUDPTimeSeqRecord,
  StripDSRHeader, StripEtherVLANHeader, TCPFragmenter, UDPIP6Encap,
  UnstripDSRHeader.

* Improved elements: ARPQuerier, AddressInfo, BandwidthRatedSplitter,
  BandwidthRatedUnqueue, CheckLength, DRRSched, FromDevice linuxmodule,
  FromDevice userlevel, FromHost userlevel, FromIPSummaryDump,
  FromSimDevice, FromUserDevice, GetIPAddress, HostEtherFilter, IPFilter
  ("ether" comparisons, faster initialization), InfiniteSource, KernelTun,
  LinkUnqueue, RandomSource, RatedSource, RatedSplitter, RatedUnqueue,
  Script, SetIPAddress, SetTimestamp, TCPIPSend, TimeSortedSched, ToDevice
  userlevel, ToUserDevice, Truncate, Unqueue2, and others.

* Element and tool bug fixes: CheckTCPHeader, CheckUDPHeader, DSRRouteTable
  DelayUnqueue, FastUDPSourceIP6, FromHost, FromUserDevice, IP6Encap,
  IPAddrRewriter, IPFilter and other classifiers, KernelFilter,
  RadiotapDecap, RadixIPLookup, SetIP6DSCP, Socket, ThreadSafeQueue,
  TimeSortedSched, ToUserDevice, UDPIPEncap, WifiDecap, click-align,
  click-undead, and others.

* Performance improvements

** Tasks: A simplified reschedule implementation and streamlined driver
   loop is both faster and less error prone (now calling fast_reschedule()
   outside of Element::run_task() is safe).

** Tasks: If a task does no work, delay re-running it.  This simple change
   improves performance on many configurations, particularly those mixing
   file descriptor or timer access and Tasks.

** Timers: A 4-heap improves performance with very large numbers of timers.

** User-level: Device elements default to libpcap even on Linux, offering
   improved performance depending on how libpcap was compiled (e.g. packet
   mmap support).  A packet memory pool suggested by Luigi Rizzo offers
   much better performance for packet allocation and freeing.

** The round-robin scheduler (--disable-stride) works again.

** New configure option --enable-bound-port-transfer slightly optimizes
   push() and pull() calls (Cliff Frey).

* Language changes

** New "require(library FILE)" allows one Click script to include another
   (Christian Kreibich).

** Many-to-one, one-to-many, and many-to-many connections cleanly express
   some connection patterns.  For example, "c :: Classifier => a1, a2, a3"
   means the same as "c :: Classifier; c [0] -> a1; c [1] -> a2; c [2] ->
   a3".

** Element groups cleanly express short detours from a connection path.
   For example, this:

	c :: Classifier(00/01);
	x -> c -> y;
	c [1] -> Paint(1) -> y;

   can be expressed like this:

	x -> c :: Classifier(00/01) => (
            [0] -> [0]; [1] -> Paint(1) -> [0]
        ) -> y;

* Multithreading

** Many performance and stability improvements to multithreaded
   configurations, particularly at userlevel.

** Lazy Task rescheduling reduces opportunities for bugs.

** Each thread gains its own sets of Timers (and, at user level, selector
   file descriptors).  An element's Timers are initially scheduled on the
   same thread as its Tasks.  This should simplify reasoning about
   Task/Timer conflicts, and can help improve performance.

* Linuxmodule

** Patchless installs are on by default.  Click installs and works without
   a patch on many versions of Linux up to 2.6.38.

* Clicky changes

** Per-router styles: ClickyInfo elements let a script specify how it
   should be drawn.

** "Run" and "Stop" buttons can execute a configuration.

* Internals

** Args: This new class and its helpers support concise, easy-to-use,
   type-safe argument parsing and are strongly preferred to cp_va_kparse in
   new code.

** TokenBucket: This new template class supports fast, flexible,
   integer-based token bucket rate management.

** Timestamps default to nanosecond precision on many platforms.

** Element::simple_action() can be used for multi-port elements.

** On some platforms, set environment variable CLICK_BACKTRACE to 1 and get
   a useful stack backtrace if Click crashes (Cliff Frey).

** Other improvements: more flexible handlers, simple_action, type traits,
   faster multi-precision arithmetic, better IPv6 address unparsing.

* Acknowledgements

** Thanks to Cliff Frey for extensive discussion, multithreaded debugging,
   802.1Q elements, TokenBucket, TCPFragmenter, bound port transfer, and
   others.

** Thanks to Joonwoo Park for Linux kernel module patches.


tools/click-install/click-install.cc
/*
 * click-install.cc -- configuration installer for Click kernel module
 * Eddie Kohler
 *
 * Copyright (c) 1999-2000 Massachusetts Institute of Technology
 * Copyright (c) 2000 Mazu Networks, Inc.
 * Copyright (c) 2002 International Computer Science Institute
 * Copyright (c) 2006 Regents of the University of California
 * Copyright (c) 2008 Meraki, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, subject to the conditions
 * listed in the Click LICENSE file. These conditions include: you must
 * preserve this copyright notice, and you cannot mention the copyright
 * holders in advertising related to the Software without their permission.
 * The Software is provided WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED. This
 * notice is a summary of the Click LICENSE file; the license in that file is
 * legally binding.
 */

#include <click/config.h>

#include "common.hh"
#include "routert.hh"
#include "lexert.hh"
#include <click/error.hh>
#include <click/args.hh>
#include <click/clp.h>
#include <click/straccum.hh>
#include <click/driver.hh>
#include "toolutils.hh"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#if FOR_BSDMODULE
# include <sys/param.h>
# include <sys/mount.h>
#elif FOR_LINUXMODULE
# include <sys/mount.h>
# include <pwd.h>
# include <grp.h>
#endif
#include <fcntl.h>
#include <unistd.h>

#define HELP_OPT		300
#define VERSION_OPT		301
#define CLICKPATH_OPT		302
#define ROUTER_OPT		303
#define EXPRESSION_OPT		304
#define UNINSTALL_OPT		305
#define HOTSWAP_OPT		306
#define MAP_OPT			307
#define VERBOSE_OPT		308
#define THREADS_OPT		309
#define PRIVATE_OPT		310
#define PRIORITY_OPT		311
#define GREEDY_OPT		312
#define UID_OPT			313
#define GID_OPT			314
#define CPU_OPT			315

static const Clp_Option options[] = {
  { "cabalistic", 0, PRIVATE_OPT, 0, Clp_Negate },
  { "clickpath", 'C', CLICKPATH_OPT, Clp_ValString, 0 },
  { "expression", 'e', EXPRESSION_OPT, Clp_ValString, 0 },
  { "file", 'f', ROUTER_OPT, Clp_ValString, 0 },
  { "help", 0, HELP_OPT, 0, 0 },
  { "hot-swap", 'h', HOTSWAP_OPT, 0, Clp_Negate },
  { "hotswap", 'h', HOTSWAP_OPT, 0, Clp_Negate },
  { "priority", 'n', PRIORITY_OPT, Clp_ValInt, 0 },
#if FOR_LINUXMODULE
  { "map", 'm', MAP_OPT, 0, 0 },
  { "private", 'p', PRIVATE_OPT, 0, Clp_Negate },
  { "threads", 'j', THREADS_OPT, Clp_ValUnsigned, 0 },
  { 0, 't', THREADS_OPT, Clp_ValUnsigned, 0 }, // deprecated
  { "greedy", 'G', GREEDY_OPT, 0, Clp_Negate },
  { "uid", 'U', UID_OPT, Clp_ValString, 0 },
  { "user", 0, UID_OPT, Clp_ValString, 0 },
  { "gid", 0, GID_OPT, Clp_ValString, 0 },
  { "cpu", 0, CPU_OPT, Clp_ValUnsigned, 0 },
#endif
  { "uninstall", 'u', UNINSTALL_OPT, 0, Clp_Negate },
  { "verbose", 'V', VERBOSE_OPT, 0, Clp_Negate },
  { "version", 'v', VERSION_OPT, 0, 0 },
};

static const char *program_name;
#if FOR_LINUXMODULE
static bool output_map;
#endif

static String tmpdir;

void
short_usage()
{
  fprintf(stderr, "Usage: %s [OPTION]... [ROUTERFILE]\n\
Try '%s --help' for more information.\n",
	  program_name, program_name);
}

void
usage()
{
  printf("\
'Click-install' installs a kernel Click configuration. It loads the Click\n\
kernel module, and any other necessary modules, as required.\n\
\n\
Usage: %s [OPTION]... [ROUTERFILE]\n\
\n\
Options:\n\
  -f, --file FILE          Read router configuration from FILE.\n\
  -e, --expression EXPR    Use EXPR as router configuration.\n\
  -h, --hot-swap           Hot-swap install new configuration.\n\
  -u, --uninstall          Uninstall Click from kernel, then reinstall.\n\
  -n, --priority N         Set kernel thread priority to N (lower is better).\n", program_name);
#if FOR_LINUXMODULE
  printf("\
  -p, --private            Make /click readable only by owning user.\n\
  -U, --user USER[:GROUP]  Set owning user [root].\n\
  -j, --threads N          Use N threads (multithreaded Click only).\n\
  -G, --greedy             Make Click thread take up an entire CPU.\n\
      --cpu N              Click thread runs on CPU N.\n");
# if HAVE_LINUXMODULE_2_6
  printf("\
  -m, --map                Print load map to the standard output.\n");
# endif
#endif
  printf("\
  -V, --verbose            Print information about files installed.\n\
  -C, --clickpath PATH     Use PATH for CLICKPATH.\n\
      --help               Print this message and exit.\n\
  -v, --version            Print version number and exit.\n\
\n\
Report bugs to <click@pdos.lcs.mit.edu>.\n");
}

static void
prepare_tmpdir(ErrorHandler *errh)
{
  ContextErrorHandler cerrh(errh, "While preparing to compile packages:");
  BailErrorHandler berrh(&cerrh);

  // change to temporary directory
  tmpdir = click_mktmpdir(&berrh);
  assert(tmpdir);
  if (chdir(tmpdir.c_str()) < 0)
    berrh.fatal("cannot chdir to %s: %s", tmpdir.c_str(), strerror(errno));
}

static void
compile_archive_packages(RouterT *r, HashTable<String, int> &packages,
			 ErrorHandler *errh)
{
  bool tmpdir_populated = false;

  // go over requirements
  Vector<String> requirements = r->requirements();
  for (int i = 0; i < requirements.size(); i += 2) {
    if (!requirements[i].equals("package", 7))
      continue;

    const String &req = requirements[i+1];

    // skip if already have object file
    if (r->archive_index(req + OBJSUFFIX) >= 0
	|| packages[req] >= 0)
      continue;

    // look for source file, prepare temporary directory
    int source_ae = r->archive_index(req + CXXSUFFIX);
    if (source_ae < 0)
      source_ae = r->archive_index(req + ".cc");
    if (source_ae < 0)
      continue;

    // found source file, so compile it
    errh->message("Compiling package %s from config archive", req.c_str());
    String result_filename = click_compile_archive_file(r->archive(), &r->archive()[source_ae], req, COMPILETARGET, false, tmpdir_populated, errh);

    // grab object file and add to archive
    if (result_filename) {
	ArchiveElement obj_ae = init_archive_element(req + OBJSUFFIX, 0600);
	obj_ae.data = file_string(result_filename, errh);
	r->add_archive(obj_ae);
    }
  }
}

static void
install_module(const String &filename, const String &options,
	       ErrorHandler *errh)
{
#if FOR_LINUXMODULE
  String cmdline = "/sbin/insmod ";
  if (output_map)
    cmdline += "-m ";
  cmdline += filename;
  if (options)
    cmdline += " " + options;
  int retval = system(cmdline.c_str());
  if (retval != 0)
    errh->fatal("'%s' failed", cmdline.c_str());
#else
  String cmdline = "/sbin/kldload " + filename;
  assert(!options);
  int retval = system(cmdline.c_str());
  if (retval != 0)
    errh->fatal("'%s' failed", cmdline.c_str());
#endif
}

static void
install_required_packages(RouterT *r, HashTable<String, int> &packages,
			  HashTable<String, int> &active_modules,
			  ErrorHandler *errh)
{
  // check for uncompiled archive packages and try to compile them
  compile_archive_packages(r, packages, errh);

  // go over requirements
  Vector<String> requirements = r->requirements();
  for (int i = 0; i < requirements.size(); i += 2) {
    if (!requirements[i].equals("package", 7))
	continue;

    String req = requirements[i+1];

    // look for object in archive
    int obj_aei = r->archive_index(req + OBJSUFFIX);
    if (obj_aei >= 0 && packages[req] < 0) {
      // install archived objects. mark them with leading underscores.
      // may require renaming to avoid clashes in 'insmod'

      // choose module name
      String insmod_name = req + OBJSUFFIX;

      if (verbose)
	errh->message("Installing package %s (%s" OBJSUFFIX " from config archive)", insmod_name.c_str(), req.c_str());

      // install module
      if (!tmpdir)
	prepare_tmpdir(errh);
      const ArchiveElement &ae = r->archive(obj_aei);
      String tmpnam = tmpdir + insmod_name;
      FILE *f = fopen(tmpnam.c_str(), "w");
      if (!f)
	errh->fatal("%s: %s", tmpnam.c_str(), strerror(errno));
      ignore_result(fwrite(ae.data.data(), 1, ae.data.length(), f));
      fclose(f);

      install_module(tmpnam, String(), errh);

      // cleanup
      packages.set(req, 1);
      active_modules.set(insmod_name, 1);

    } else if (packages[req] < 0) {
      // install required package from CLICKPATH
      String filename = req + OBJSUFFIX;
      String pathname = clickpath_find_file(filename, "lib", CLICK_LIBDIR);
      if (!pathname) {
	filename = req + ".o";
	pathname = clickpath_find_file(filename, "lib", CLICK_LIBDIR);
	if (!pathname)
	  errh->fatal("cannot find required package '%s" OBJSUFFIX "'\nin CLICKPATH or '%s'", req.c_str(), CLICK_LIBDIR);
      }

      // install module
      if (verbose)
	errh->message("Installing package %s (%s)", req.c_str(), pathname.c_str());

      install_module(pathname, String(), errh);

      packages.set(req, 1);
      active_modules.set(filename, 1);

    } else {
      // package already loaded; note in 'active_modules' that we still need
      // it
      if (verbose)
	errh->message("Not installing package %s, version already exists", req.c_str());

      String filename = req;
      if (active_modules[filename] < 0)
	filename = req + OBJSUFFIX;
      if (active_modules[filename] < 0)
	filename = req + ".o";
      if (active_modules[filename] == 0)
	active_modules.set(filename, 1);
    }
  }
}

int
main(int argc, char **argv)
{
  click_static_initialize();
  CLICK_DEFAULT_PROVIDES;
  ErrorHandler *nop_errh = ErrorHandler::default_handler();
  ErrorHandler *errh = new PrefixErrorHandler(nop_errh, "click-install: ");

  // read command line arguments
  Clp_Parser *clp =
    Clp_NewParser(argc, argv, sizeof(options) / sizeof(options[0]), options);
  Clp_SetOptionChar(clp, '+', Clp_ShortNegated);
  program_name = Clp_ProgramName(clp);

  const char *router_file = 0;
  bool file_is_expr = false;
  bool uninstall = false;
  bool hotswap = false;
  int priority = -100;
#if FOR_LINUXMODULE
  bool accessible = true;
  int threads = 1;
  bool greedy = false;
  output_map = false;
  uid_t uid = 0;
  gid_t gid = 0;
  int cpu = -1;
#endif

  while (1) {
    int opt = Clp_Next(clp);
    switch (opt) {

     case HELP_OPT:
      usage();
      exit(0);
      break;

     case VERSION_OPT:
      printf("click-install (Click) %s\n", CLICK_VERSION);
      printf("Click packages in %s, binaries in %s\n", CLICK_LIBDIR, CLICK_BINDIR);
      printf("Copyright (c) 1999-2000 Massachusetts Institute of Technology\n\
Copyright (c) 2000-2005 Mazu Networks, Inc.\n\
Copyright (c) 2002 International Computer Science Institute\n\
This is free software; see the source for copying conditions.\n\
There is NO warranty, not even for merchantability or fitness for a\n\
particular purpose.\n");
      exit(0);
      break;

     case CLICKPATH_OPT:
      set_clickpath(clp->vstr);
      break;

     case ROUTER_OPT:
     case EXPRESSION_OPT:
     router_file:
      if (router_file) {
	errh->error("router configuration specified twice");
	goto bad_option;
      }
      router_file = clp->vstr;
      file_is_expr = (opt == EXPRESSION_OPT);
      break;

     case Clp_NotOption:
      if (!click_maybe_define(clp->vstr, errh))
	  goto router_file;
      break;

#if FOR_LINUXMODULE
     case THREADS_OPT:
      threads = clp->val.u;
      if (threads < 1) {
        errh->error("must have at least one thread");
	goto bad_option;
      }
      break;

     case PRIVATE_OPT:
      accessible = clp->negated;
      break;

     case PRIORITY_OPT:
      priority = clp->val.i;
      break;

    case MAP_OPT:
# if HAVE_LINUXMODULE_2_6
	errh->warning("'%s' ignored on 2.6 kernels", Clp_CurOptionName(clp));
# else
	output_map = !clp->negated;
# endif
	break;

    case GREEDY_OPT:
	greedy = !clp->negated;
	break;

    case UID_OPT: {
	const char *colon = find(clp->vstr, clp->vstr + strlen(clp->vstr), ':');
	if (colon > clp->vstr) {
	    String s(clp->vstr, colon);
	    if (!IntArg().parse(s, uid)) {
		errno = 0;
		struct passwd *pwd = getpwnam(s.c_str());
		if (!pwd && errno)
		    errh->error("username lookup: %s", strerror(errno));
		else if (!pwd)
		    errh->error("no such user '%s'", s.c_str());
		else
		    uid = pwd->pw_uid;
	    }
	}
	if (*colon && colon[1]) {
	    clp->vstr = colon + 1;
	    goto gid;
	}
	break;
    }

    gid:
    case GID_OPT: {
	if (!IntArg().parse(clp->vstr, gid)) {
	    errno = 0;
	    struct group *grp = getgrnam(clp->vstr);
	    if (!grp && errno)
		errh->error("group lookup: %s", strerror(errno));
	    else if (!grp)
		errh->error("no such group '%s'", clp->vstr);
	    else
		gid = grp->gr_gid;
	}
	break;
    }

     case CPU_OPT:
      cpu = clp->val.i;
      break;
#endif

     case UNINSTALL_OPT:
      uninstall = !clp->negated;
      break;

     case HOTSWAP_OPT:
      hotswap = !clp->negated;
      break;

     case VERBOSE_OPT:
      verbose = !clp->negated;
      break;

     bad_option:
     case Clp_BadOption:
      short_usage();
      exit(1);
      break;

     case Clp_Done:
      goto done;

    }
  }

 done:
  // check options
  if (hotswap && uninstall)
    errh->warning("'--hotswap' and '--uninstall' are mutually exclusive");

  RouterT *r = read_router(router_file, file_is_expr, nop_errh);
  if (r)
    r->flatten(nop_errh, true);
  if (!r || errh->nerrors() > 0)
    exit(1);

  // pathnames of important Click files
  String clickfs_packages = clickfs_prefix + String("/packages");

  // uninstall Click if requested
  if (uninstall)
    unload_click(errh);

  // install Click module if required
  if (access(clickfs_packages.c_str(), F_OK) < 0) {
#if FOR_LINUXMODULE
    // find and install proclikefs.o
    StringMap modules(-1);
    if (read_active_modules(modules, errh) && modules["proclikefs"] < 0) {
# if HAVE_LINUXMODULE_2_6
      String proclikefs_o =
	clickpath_find_file("proclikefs.ko", "lib", CLICK_LIBDIR, errh);
# else
      String proclikefs_o =
	clickpath_find_file("proclikefs.o", "lib", CLICK_LIBDIR, errh);
# endif
      if (verbose)
	errh->message("Installing proclikefs (%s)", proclikefs_o.c_str());
      install_module(proclikefs_o, String(), errh);
    }
#endif

    // find loadable module
#if FOR_BSDMODULE || (FOR_LINUXMODULE && HAVE_LINUXMODULE_2_6)
    String click_o =
      clickpath_find_file("click.ko", "lib", CLICK_LIBDIR, errh);
#elif FOR_LINUXMODULE
    String click_o =
      clickpath_find_file("click.o", "lib", CLICK_LIBDIR, errh);
#endif
    if (verbose)
      errh->message("Installing Click module (%s)", click_o.c_str());

    // install it in the kernel
#if FOR_LINUXMODULE
    String options;
    if (threads > 1)
      options += "threads=" + String(threads);
    if (greedy)
	options += " greedy=1";
    if (!accessible)
	options += " accessible=0";
    if (uid != 0)
	options += " uid=" + String(uid);
    if (gid != 0)
	options += " gid=" + String(gid);
    if (cpu != -1)
	options += " cpu=" + String(cpu);
    install_module(click_o, options, errh);
#elif FOR_BSDMODULE
    install_module(click_o, String(), errh);
#endif

#if FOR_BSDMODULE || FOR_LINUXMODULE
    // make clickfs_prefix directory if required
    if (access(clickfs_dir.c_str(), F_OK) < 0 && errno == ENOENT) {
      if (mkdir(clickfs_dir.c_str(), 0777) < 0)
	errh->fatal("cannot make directory %s: %s", clickfs_dir.c_str(), strerror(errno));
    }

    // mount Click file system
    if (verbose)
      errh->message("Mounting Click module at %s", clickfs_dir.c_str());
# if FOR_BSDMODULE
    String cmdline = "mount -t click click ";
    cmdline += clickfs_dir.c_str();
    int mount_retval = system(cmdline.c_str());
# else
    int mount_retval = mount("none", clickfs_dir.c_str(), "click", 0, 0);
# endif
    if (mount_retval < 0 && (verbose || errno != EBUSY))
      errh->error("cannot mount %s: %s", clickfs_dir.c_str(), strerror(errno));
#endif

    // check that all is well
    if (access(clickfs_packages.c_str(), F_OK) < 0)
      errh->fatal("cannot install Click module");
  } else {
#if FOR_LINUXMODULE
    if (threads > 1 || greedy || !accessible || uid != 0 || gid != 0 || cpu != -1)
      errh->warning("Click module already installed, some options ignored");
#endif
  }

  // check for new-style directory layout
  if (adjust_clickfs_prefix())
      clickfs_packages = clickfs_prefix + String("/packages");

  String clickfs_config = clickfs_prefix + String("/config");
  String clickfs_hotconfig = clickfs_prefix + String("/hotconfig");
  String clickfs_errors = clickfs_prefix + String("/errors");
  String clickfs_priority = clickfs_prefix + String("/priority");

  // find current packages
  HashTable<String, int> active_modules(-1);
  HashTable<String, int> packages(-1);
  read_active_modules(active_modules, errh);
  read_package_file(clickfs_packages, packages, errh);

  // install required packages
  install_required_packages(r, packages, active_modules, errh);

  // set priority
  if (priority > -100) {
    FILE *f = fopen(clickfs_priority.c_str(), "w");
    if (!f)
      errh->fatal("%s: %s", clickfs_priority.c_str(), strerror(errno));
    fprintf(f, "%d\n", priority);
    fclose(f);
  }

  // write flattened configuration to CLICKFS/config
  int exit_status = 0;
  {
    String config_place = (hotswap ? clickfs_hotconfig : clickfs_config);
    if (verbose)
      errh->message("Writing configuration to %s", config_place.c_str());
    int fd = open(config_place.c_str(), O_WRONLY | O_TRUNC);
    if (fd < 0)
      errh->fatal("cannot install configuration: %s", strerror(errno));
    // XXX include packages?
    String config = r->configuration_string();
    int pos = 0;
    while (pos < config.length()) {
      ssize_t written = write(fd, config.data() + pos, config.length() - pos);
      if (written >= 0)
	pos += written;
      else if (errno != EAGAIN && errno != EINTR)
	errh->fatal("%s: %s", config_place.c_str(), strerror(errno));
    }
    int retval = close(fd);
    if (retval < 0 && errno == EINVAL)
      exit_status = 2;
    else if (retval < 0)
      errh->error("%s: %s", config_place.c_str(), strerror(errno));
  }

  // report errors
  {
    char buf[1024];
    int fd = open(clickfs_errors.c_str(), O_RDONLY | O_NONBLOCK);
    if (fd < 0)
      errh->warning("%s: %s", clickfs_errors.c_str(), strerror(errno));
    else {
      if (verbose)
	errh->message("Waiting for errors");
      while (1) {
	struct timeval wait;
	wait.tv_sec = 0;
	wait.tv_usec = 50000;
	(void) select(0, 0, 0, 0, &wait);
	ssize_t got = read(fd, buf, 1024);
	if (got > 0)
	  ignore_result(fwrite(buf, 1, got, stderr));
	else if (got == 0)
	  break;
	else if (errno != EINTR && errno != EAGAIN) {
	  errh->error("%s: %s", clickfs_errors.c_str(), strerror(errno));
	  break;
	}
      }
      close(fd);
    }
  }

  // remove unused packages
  remove_unneeded_packages(active_modules, packages, errh);

  if (verbose)
    errh->message("Done");
  exit(exit_status);
}
